tools/indexer/cmd/start.go

```diff
@@ -0,0 +1,336 @@
+
+		latestBlock, err := ethereum.GetLatestBlockNumber(httpc.HTTPClient, infuraRPC)
+		if err != nil {
+			logger.Error("failed to get latest block from RPC", "error", err)
+			return 0, err
+		}
+
+		lastBN = latestBlock - 10 // Start 10 blocks behind to ensure data availability
+		logger.Info("starting from block", "block", lastBN, "latest", latestBlock)
+	}
+	return lastBN, nil
+}
+
+func runBackfillIfConfigured(ctx context.Context, c *cli.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, logger *slog.Logger) {
+	logger.Info("indexer configuration", "lookback", c.Int("backfill-lookback"), "batch", c.Int("backfill-batch"))
+
+	if c.Int("backfill-lookback") > 0 {
+		logger.Info("[BACKFILL] running one-time backfill",
+			"lookback", c.Int("backfill-lookback"),
+			"batch", c.Int("backfill-batch"))
+		if err := backfill.RunAll(ctx, db, httpc, createOptionsFromCLI(c), relays); err != nil {
+			logger.Error("[BACKFILL] failed", "error", err)
+		} else {
+			logger.Info("[BACKFILL] completed startup backfill")
+		}
+	} else {
+		logger.Info("[BACKFILL] skipped", "reason", "backfill-lookback=0")
+	}
+}
+
+func runMainLoop(ctx context.Context, c *cli.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, infuraRPC, beaconBase string, startBN int64, logger *slog.Logger) error {
+	mainTicker := time.NewTicker(c.Duration("block-interval"))
+	defer mainTicker.Stop()
+
+	lastBN := startBN
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Info("[SHUTDOWN] graceful shutdown initiated", "reason", ctx.Err())
+			if err := db.SaveLastBlockNumber(ctx, lastBN); err != nil {
+				logger.Error("[SHUTDOWN] failed to save last block number", "error", err)
+			}
+			logger.Info("[SHUTDOWN] indexer stopped", "block", lastBN)
+			return nil
+
+		case <-mainTicker.C:
+			lastBN = processNextBlock(ctx, c, db, httpc, relays, infuraRPC, beaconBase, lastBN, logger)
+		}
+	}
+}
+
+func processNextBlock(ctx context.Context, c *cli.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, infuraRPC, beaconBase string, lastBN int64, logger *slog.Logger) int64 {
+	nextBN := lastBN + 1
+
+	ei, err := beacon.FetchCombinedBlockData(ctx, httpc, infuraRPC, beaconBase, nextBN)
+	if err != nil || ei == nil {
+		logger.Warn("[BLOCK] not available yet", "block", nextBN, "error", err)
+		return lastBN
+	}
+
+	logger.Info("processing block",
+		"block", nextBN,
+		"slot", ei.Slot,
+		"timestamp", ei.Timestamp,
+		"proposer_index", ei.ProposerIdx,
+		"winning_relay", ei.RelayTag,
+		"builder_pubkey_prefix", ei.BuilderHex,
+		"producer_reward_eth", ei.RewardEth,
+	)
+
+	if err := db.UpsertBlockFromExec(ctx, ei); err != nil {
+		logger.Error("[DB] failed to save block", "block", nextBN, "error", err)
+		return lastBN
+	}
+	logger.Info("[DB] block saved successfully", "block", nextBN)
+
+	processBidsForBlock(ctx, db, httpc, relays, ei, logger)
+	launchAsyncValidatorTasks(ctx, c, db, httpc, ei, beaconBase, logger)
```

Comment:
We launch async tasks and save block progress? What if there was an error in the async task?